### PR TITLE
refactoring: remove unnecessary allocation

### PR DESCRIPTION
**Pull request by k-morozov**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by k-morozov** [comment]
There are 2 memory allocation when we create std::shared_ptr<ContextAccess> - for  ContextAccess and control block. Althought we don't use weak_ptr I suppouse that in this case there is no need 2 memory allocation (make_shared has 1 alloc). make_shared has problem with private ctor - that's why I append static make.

**Comment by rschu1ze** [comment]
Your goal (allocating 1x instead of 2x) is a good one and the implementation is also reasonable.

I did a [big PR](https://github.com/ClickHouse/ClickHouse/pull/36438/files#diff-367517929e1287b00c2c8182a22ef93ba9f37735f0bbad4d5b15be45aa2483fb) which aimed for exactly the same. After some discussion in that other PR, it turned out that the simplest solution to solve the make_shared-doesn't-work-for-private-constructors problem is to simply make the constructor public. That crates the least amount of boilerplate and after all, constructors are supposed to be public (in the general case). Would that also be an option here?

**Comment by k-morozov** [comment]
@rschu1ze rschu1ze I agree with you. I will update this pr.

File: src/Access/AccessControl.cpp [hunk 1]
```diff
@@ -79,7 +79,7 @@ class AccessControl::ContextAccessCache
             /// No user, probably the user has been dropped while it was in the cache.
             cache.remove(params);
         }
-        auto res = std::shared_ptr<ContextAccess>(new ContextAccess(access_control, params));
+        auto res = ContextAccess::make(access_control, params);
```
Comment (rschu1ze) [inline src/Access/AccessControl.cpp:82]: A plain and simple `auto res = std::make_shared<ContextAccess>(access_control, params);` would have done the job but I am also fine with `::make()`.

**Review by rschu1ze — APPROVED** [review]

**Pull request merged**